### PR TITLE
JAMES-3087 Reactive mailbox provisionning

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
## Before 

Capture d’écran de 2020-03-22 21-57-58

## After

Capture d’écran de 2020-03-22 21-57-49

## Conclusion

 - Stable average execution time
 - P99 reduction ( 30% ), standard deviation reduction (from 102 to 62)

(Thus of course p50 is 1.6% solwer)

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 1]
```diff
@@ -129,27 +131,24 @@ private JPAMailbox jpaMailbox(Mailbox mailbox) throws MailboxException {
...
                 .setParameter("userParam", mailboxPath.getUser().asString())
                 .getSingleResult()
-                .toMailbox();
+                .toMailbox());
         } catch (NoResultException e) {
-            throw new MailboxNotFoundException(mailboxPath);
+            return Mono.empty();
         } catch (PersistenceException e) {
-            throw new MailboxException("Search of mailbox " + mailboxPath + " failed", e);
+            return Mono.error(e);
```
Comment (trantienduchn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: you don't wrap it in `MailboxException` ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java [hunk 2]
```diff
@@ -59,8 +61,21 @@
     /**
      * Return the {@link Mailbox} for the given name
      */
-    Mailbox findMailboxByPath(MailboxPath mailboxName)
-            throws MailboxException, MailboxNotFoundException;
+    Mono<Mailbox> findMailboxByPath(MailboxPath mailboxName);
```
Comment (trantienduchn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java]: `Publisher<Mailbox>` ?

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java [hunk 3]
```diff
@@ -129,27 +131,24 @@ private JPAMailbox jpaMailbox(Mailbox mailbox) throws MailboxException {
...
-            return false;
-        }
+        return findMailboxByPath(mailboxPath)
+            .blockOptional()
+            .isPresent();
     }
 
     @Override
-    public Mailbox findMailboxByPath(MailboxPath mailboxPath) throws MailboxException, MailboxNotFoundException {
+    public Mono<Mailbox> findMailboxByPath(MailboxPath mailboxPath)  {
```
Comment (trantienduchn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: can this method be done in a reactive fashion?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java]: No IMO this is withing the mailbox module thus we can keep the Mono here as this is an internal API

(internal within mailboxManager)

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMailboxMapper.java]: No I don't think so.

I'd rather keep the behaviour unchanged here...

**Comment by chibenwa** [comment]
```
[be644a2ae69442b499036a99826b1fde05044826] + docker inspect -f {{(index .NetworkSettings.IPAddress)}} james-server-cassandra-rabbitmq-reactive-mailbox-provisionning-c889fb49-eac5-44d4-b2f1-fb1644b4f68e
[be644a2ae69442b499036a99826b1fde05044826] + docker run -a stdout -a stderr --volume m2-reactive-mailbox-provisionning-c889fb49-eac5-44d4-b2f1-fb1644b4f68e:/root/.m2 --name=integration-cassandra-rabbitmq-reactive-mailbox-provisionning-c889fb49-eac5-44d4-b2f1-fb1644b4f68e --env MVN_ADDITIONAL_ARG_LINE=--batch-mode --entrypoint=/root/integration_tests.sh --volumes-from=git-publish-reactive-mailbox-provisionning-c889fb49-eac5-44d4-b2f1-fb1644b4f68e james-parent-compile-reactive-mailbox-provisionning-c889fb49-eac5-44d4-b2f1-fb1644b4f68e 172.17.0.9 143 587 be644a2ae69442b499036a99826b1fde05044826
[be644a2ae69442b499036a99826b1fde05044826] docker: Error response from daemon: No such container: git-publish-reactive-mailbox-provisionning-c889fb49-eac5-44d4-b2f1-fb1644b4f68e.
```

**Comment by chibenwa** [comment]
`MemorySpamAssassinContractTest.imapCopiesToSpamMailboxShouldBeConsideredAsSpam » ConditionTimeout`

test this please

**Comment by Arsnael** [comment]
```
[2440a3e6ed2c3aec4e02df78ce09073e78f454dd] [ERROR] Failed to execute goal on project james-server-lifecycle-api: Could not resolve dependencies for project org.apache.james:james-server-lifecycle-api:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> com.google.errorprone:error_prone_annotations:jar:2.1.3: Failed to read artifact descriptor for com.google.errorprone:error_prone_annotations:jar:2.1.3: Could not transfer artifact com.google.errorprone:error_prone_annotations:pom:2.1.3 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/com/google/errorprone/error_prone_annotations/2.1.3/error_prone_annotations-2.1.3.pom.part (No such file or directory) -> [Help 1]
```
test this please

**Comment by chibenwa** [comment]
Let's unpoc this ^^

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 4]
```diff
@@ -1800,7 +1800,7 @@ void user2ShouldBeAbleToCreateNestedFoldersWithoutTheirParents() throws Exceptio
             MailboxPath nestedFolder = MailboxPath.forUser(USER_2, "INBOX.testfolder");
             mailboxManager.createMailbox(nestedFolder, session);
 
-            assertThat(mailboxManager.mailboxExists(nestedFolder, session)).isTrue();
+            assertThat(Mono.from(mailboxManager.mailboxExists(nestedFolder, session)).block()).isTrue();
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Looks like a good idea

But such a thing out-range the responsibility of this humble PR don't you think?

[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: oh yes, of course, I don't expect that we do that in this PR

**Comment by chibenwa** [comment]
TODO of course add `JAMES-3087` in commit messages upon merge

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**